### PR TITLE
[4.0] Introduce category factory

### PR DESCRIPTION
**Pull request by laoneo**
As discussed on skype, moves the category instance creation code into a factory.

I'v also extended it with a new `CategoryInterface` class so we do not have to deal with the implementation class directly.

### CODE REVIEW

**Comment by laoneo** [comment]
I need to get this one merged to finish the service conversion of com_contact, com_newsfeeds and com_categories.

**Comment by infograf768** [comment]
I have tested this item :white_check_mark: successfully on 582fa3e8b248aad226d5ff3bb451f73d877003e9<br /><br />This PR here does not break more than is already broken (specially when concerning multilang routing in frontend (info-block).<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20693">issues.joomla.org/tracker/joomla-cms/20693</a>.</sub>

**Comment by Hackwar** [comment]
I don't really want to name this CategoriesInterface. The whole category system is a tree implementation, which is why I implemented a node interface here: #21327 I don't know how to name it in the tree-object-sense, but maybe use something that more aligns with tree terminology, instead of focusing heavily on the categories stuff. In an ideal world, JMenu would implement the same interface and menu items become a NodeInterface object, too.

**Comment by laoneo** [comment]
I agree for the tree, but the object which returns the tree is category related. So it makes sense to name it `CategoryInterface`. The implementation class should probably being renamed.

File: libraries/src/Categories/Categories.php [hunk 1]
```diff
@@ -160,16 +160,18 @@ public static function getInstance($extension, $options = array())
...
 	 *
-	 * @return  CategoryNode|null|boolean  CategoryNode object or null if $id is not valid
+	 * @return  CategoryNode  CategoryNode object
 	 *
 	 * @since   1.6
+	 *
+	 * @throws  CategoryNotFoundException
 	 */
-	public function get($id = 'root', $forceload = false)
+	public function get($id = 'root', $forceload = false): CategoryNode
```
Comment (wilsonge) [inline libraries/src/Categories/Categories.php]: Technically this is a b/c break for people who are subclassing JCategories and overriding this method.

Having said that I doubt people are overriding this method so it's probably ok. any thoughts?

File: libraries/src/Categories/CategoryInterface.php [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+	 * Allows to set some optional options, eg. if the access level should be considered.
+	 * Also clears the internal children cache.
+	 *
+	 * @param   array  $options  The new options
+	 *
+	 * @return  void
+	 *
+	 * @since  __DEPLOY_VERSION__
+	 */
+	public function setOptions(array $options);
```
Comment (wilsonge) [inline libraries/src/Categories/CategoryInterface.php]: Do we really want this in the interface? I think the options setting is specific to the implementation?

File: libraries/src/Categories/Categories.php (hunk 1 above)
Comment (laoneo) [reply libraries/src/Categories/Categories.php]: Agree, lets revert it then.

File: libraries/src/Categories/CategoryInterface.php (hunk 2 above)
Comment (laoneo) [reply libraries/src/Categories/CategoryInterface.php]: In the router you get a Category Injected and then you want to fetch the children with access check and without, so we need it there, or we inject there then the factory.

**Comment by laoneo** [comment]
Can we get this merged @wilsonge so I can finish the service provider story?

**Comment by wilsonge** [comment]
I know it's annoyingly retrospective. But this looks good to me.

**Pull request merged**